<a href="https://colab.research.google.com/github/jakubdaron/machine-learning-course-part-1/blob/main/Regresja/01_linear_regression_normal_education.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import bibliotek

In [2]:
import numpy as np

Przygotowanie danych

In [3]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [4]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [5]:
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [6]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


Równanie normalne

równanie pozwalajace obliczyc minimum funkcji straty:
- W = (X^T * X)^-1 * (X^T * Y)

In [7]:
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [8]:
L = np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [9]:
P = np.dot(X.T, Y)
P

array([21750., 80500.])

In [10]:
np.dot(L, P)

array([2750.,  250.])

Końcowa postać modelu:
- Y = 2500 + 250 X1

Regresja liniowa przy pomocy scikit-learn

In [11]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
